raysect/core/math/function/float/function1d/interpolate.pyx

```diff
@@ -0,0 +1,538 @@
+        self._f = f
+        self._last_index = self._x.shape[0] -1
+        self._extrapolation_range = extrapolation_range
+
+        # dimensions checks
+        if x.ndim != 1:
+            raise ValueError(f'The x array must be 1D. Got {x.shape}.')
+
+        if f.ndim != 1:
+            raise ValueError(f'The x array must be 1D. Got {f.shape}.')
+
+        if x.shape != f.shape:
+            raise ValueError(f'Shape mismatch between x array ({x.shape}) and f array ({f.shape}).')
+
+        # test monotonicity
+        if (np.diff(x) <= 0).any():
+            raise ValueError('The x array must be monotonically increasing.')
+
+        # create extrapolator per extrapolation_type argument
+        if extrapolator_idx == EXT_NONE:
+            self._extrapolator = _ExtrapolatorNone(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_NEAREST:
+            self._extrapolator = _Extrapolator1DNearest(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_LINEAR:
+            self._extrapolator = _Extrapolator1DLinear(x, f, extrapolation_range)
+        elif extrapolator_idx == EXT_QUADRATIC:
+            raise NotImplementedError("Quadratic extrapolator not implemented.")
+        else:
+            raise ValueError(f'Unsupported extrapolator type {extrapolation_type}.')
+
+        # create extrapolator per interpolation_type argument
+        if interpolator_idx == INT_LINEAR:
+            self._interpolator = _Interpolator1DLinear(self._x, self._f)
+        elif interpolator_idx == INT_CUBIC:
+            self._interpolator = _Interpolator1DCubic(self._x, self._f)
+        # elif interpolator_idx == INT_CUBIC_CONSTRAINED:
+        #     raise NotImplementedError("Constrained cubic interpolator not implemented.")
+        else:
+            raise ValueError(f'Interpolation type {interpolation_type} not supported.')
+
+    cdef double evaluate(self, double x) except? -1e999:
+        """
+        Evaluates the interpolating function.
+
+        :param double x: the point for which an interpolated value is required
+        :return: the interpolated value at point x.
+        """
+        cdef int index = find_index(self._x, x)
+
+        if index == -1:
+            if x < self._x[0] - self._extrapolation_range:
+                raise ValueError(
+                    f'The specified value (x={x}) is outside of extrapolation range.')
+            return self._extrapolator.evaluate(x, index)
+        elif index == self._last_index:
+            if x > self._x[self._last_index] + self._extrapolation_range:
+                raise ValueError(
+                    f'The specified value (x={x}) is outside of extrapolation range.')
+            return self._extrapolator.evaluate(x, index)
+        else:
+            return self._interpolator.evaluate(x, index)
+
+    @property
+    def domain(self):
+        """
+        Returns bounding box of the provided inputs.
+        Order: min(x), max(x), min(f), max(f)
```

Comment:
The min/max f is not required (and is not even valid for cubic interpolation). This is intended to be the sampling domain, so just the function input variable ranges.

In higher dimensions you'll want to group these in tuples, otherwise this will become a mess for higher dimensions objects. e.g. for 3D; ((xmin, xmax), (ymin, ymax), (zmin, zmax)). But for 1D, just (xmin, xmax) will do.